src/sisl/cli/sdata.py

```diff
@@ -0,0 +1,370 @@
+ the most likely scenario. In that case, all the modifications required
+ are OUTSIDE of this file. You need to:
+
+    1. Create a function that generates the data from the file. Add type
+    annotations to it. E.g.:
+
+    ```python
+    import DOSData
+
+    def dos_from_bandsSile(
+        sile: sisl.io.bandsSileSiesta,
+        Erange: Tuple[float, float] = (-2, 2)
+    ):
+        # Get energies and DOS somehow
+        ...
+        E = ...
+        DOS = ...
+
+        return DOSData.new(E, DOS)
+    ```
+
+    2. Register your function as a way of creating DOSData:
+    ```python
+    DOSData.new.register("siesta-bands", dos_from_bandsSile)
+    ```
+
+    3. This data generating function should be automatically added to the CLI as
+    'siesta-bands'. Enjoy!
+
+  - By adding a post processing method to a data class. In this case, all you need to
+    do is to add the method to the data class and make sure it is type annotated. The
+    method should then automatically show up in the CLI.
+
+  - By adding a new type of data. In this case, you need to create the data class and
+    make sure that it is added to the CLI in this file. You need to add a new line to
+    the `main` function that calls `add_data_group` with the new data class.
+
+Alternatively, it is also possible that you might want to change the way the CLI works.
+This is of course more complex, but here is a sketch of how the CLI works to help
+pinpoint where you might want to make changes:
+
+    - The CLI is a click CLI.
+    - The CLI is created on `main`.
+    - The main app is created and then we add a group for each data class. Each group
+      contains:
+        - Commands to generate data:
+          We automatically gather all the possible sources from the `new` dispatcher
+          of the data class. We then decorate them automatically with
+          `decorate_data_gen_function` to be used in click based on type annotations.
+        - Commands to post process data:
+          We automatically gather all the methods of the data class that are not private
+          and decorate them with `decorate_data_method` to be used in click based on type
+          annotations.
+
+    - The way we can generate and post process data in the same CLI is by allowing commands
+      to be chained. However, since commands execute independently, we need a global storage
+      so that the post processing commands can access the generated data. We store the
+      generated data in the `GENERATED` global variable.
+
+    - To decorate functions automatically based on type, we rely on the `decorate_click`
+      function from `nodify`. However, there are some types that are specific to `sisl`. We
+      need to define how those types will be generated from the values (strings) passed in
+      the `CLI`. We do that in the `custom_type_kwargs` dictionary.
+
+"""
+import functools
+import inspect
+import sys
+from typing import Callable, Optional
+
+import rich_click as click
```

Comment:
I am probably more in favor of importing click, and if available importing rich-click.